### PR TITLE
implement onRequestMetadata/onResponseMetadata

### PR DESCRIPTION
**Pull request by zirain**
Hi @mathetake , I'm new to `proxy-wasm-go-sdk` and trying to fixed https://github.com/tetratelabs/proxy-wasm-go-sdk/issues/91 .

### CODE REVIEW

**Comment by mathetake** [comment]
thanks! Would it be possible to add e2e tests?

File: proxywasm/types/context.go [hunk 1]
```diff
@@ -145,6 +149,10 @@ type HttpContext interface {
 	// Return types.ActionPause if you want to stop sending headers to downstream.
 	OnHttpResponseHeaders(numHeaders int, endOfStream bool) Action
 
+	// OnHttpResponseMetadata is called when response metadata arrive.
```
Comment (mathetake) [inline proxywasm/types/context.go:152]:
```suggestion
	// OnHttpResponseMetadata is called when response metadata arrives.
```

File: proxywasm/types/context.go [hunk 2]
```diff
@@ -131,6 +131,10 @@ type HttpContext interface {
 	// Return types.ActionPause if you want to stop sending headers to the upstream.
 	OnHttpRequestHeaders(numHeaders int, endOfStream bool) Action
 
+	// OnHttpRequestMetadata is called when request metadata arrive.
```
Comment (mathetake) [inline proxywasm/types/context.go:134]:
```suggestion
	// OnHttpRequestMetadata is called when request metadata arrives.
```

**Comment by zirain** [comment]
> thanks! Would it be possible to add e2e tests?

I think it's ok, but need a little bit time.

I'm trying to figure out why always fail on my env when run e2e test
```console
go test -v ./e2e -count=1
=== RUN   Test_dispatch_call_on_tick
    main_test.go:63: 
                Error Trace:    main_test.go:63
                                                        e2e_test.go:30
                Error:          Condition never satisfied
                Test:           Test_dispatch_call_on_tick
                Messages:       Envoy has not started
--- FAIL: Test_dispatch_call_on_tick (5.00s)
```

```console
root@istiodev /m/d/g/s/i/proxy-wasm-go-sdk (main) [SIGINT]# which envoy
/usr/local/bin/envoy
```

**Comment by mathetake** [comment]
maybe you have to do `make build.examples` before running e2e as it uses example binaries.  https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/Makefile#L9-L10

**Comment by zirain** [comment]
> maybe you have to do `make build.examples` before running e2e as it uses example binaries. https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/Makefile#L9-L10

that works fine, maybe we need update doc or makefile?

**Comment by mathetake** [comment]
I would appreciate if you could fix makefile and add file dependency on `make tset.e2e`

**Comment by mathetake** [comment]
what I want you to add is not only an example but the actual test running on envoy: https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/e2e/e2e_test.go

File: examples/http_metadata/main.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	for _, h := range hs {
+		proxywasm.LogInfof("request header --> %s: %s", h[0], h[1])
+	}
+	return types.ActionContinue
+}
+
+// Override types.DefaultHttpContext.
+func (ctx *httpMetadata) OnHttpStreamDone() {
+	proxywasm.LogInfof("%d finished", ctx.contextID)
+}
```
Comment (mathetake) [inline examples/http_metadata/main.go:88]: we could remove this as this is shown in another example
```suggestion
```

File: examples/http_metadata/main.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	hs, err := proxywasm.GetHttpRequestHeaders()
+	if err != nil {
+		proxywasm.LogCriticalf("failed to get request headers: %v", err)
+	}
+
+	for _, h := range hs {
+		proxywasm.LogInfof("request header --> %s: %s", h[0], h[1])
+	}
+	return types.ActionContinue
+}
```
Comment (mathetake) [inline examples/http_metadata/main.go:83]: This can be deleted as well
```suggestion
```

**Comment by zirain** [comment]
> what I want you to add is not only an example but the actual test running on envoy: https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/main/e2e/e2e_test.go

yes, but I got no idea about how to simulate a http request will trigger `OnHttpRequestMetadata`, any suggestions?

**Comment by mathetake** [comment]
maybe speaking HTTP text directly via tcp connection?

**Comment by zirain** [comment]
> maybe speaking HTTP text directly via tcp connection?

let me try with that.

**Comment by zirain** [comment]
failed to setup a demo, @mathetake can you share how to do that when you're free, or you can take over the whole PR.

**Pull request closed without merging**